### PR TITLE
Palette optimization

### PR DESCRIPTION
**Pull request by AidanMars2**
## Proposed changes

`fill` has a new overload that partially fills the palette with a cuboid.

This PR optimizes several functions by using the new `PaletteIndexMap`:
- `setAll` and `replaceAll` don't immediately default to direct palette anymore.
- `optimize` now makes a new index map directly and returns early if no optimization is possible.

Fixed bugs:
- `copyFrom` now handles negative offsets properly.
- `replace(int oldValue, int newValue)` now properly replaces values if the new value already exists in the palette.
- High values are now properly accounted for in direct palettes.

Miscellaneous optimizations:
- `copyFrom` now defaults to `fill` if the source palette is single value or filled with air.
- `Palettes#remap` now only allocates if forced to.
- `load` now calls `remap` instead of manually setting all values.

Breaking changes:
- `copyFrom` now throws an exception if offset is out of bounds.
- `Palettes` is now marked as internal and some unnecessary methods were removed.
- `Palette#BLOCK_PALETTE_DIRECT_BITS` is now marked as internal.
- `Palette#offset(int)` has been removed.
- `Palette#height` now takes an `IntPredicate` instead of an `EntryPredicate`

## Types of changes

- [x] Bugfix (non-breaking change which fixes an issue)
- [x] New feature (non-breaking change which adds functionality)
- [x] Breaking change (fix or feature that would cause existing functionality to not work as expected)

## Checklist

- [x] I have read the [CONTRIBUTING.md](CONTRIBUTING.md)
- [x] I have added tests that prove my fix is effective or that my feature works
- [x] I have added necessary documentation (if appropriate)

## Further comments

The old system of using an `Int2IntMap` and `IntArrayList` was a mess, using 5 lines of code just to initialize the palette with a single value. 
This led to several bugs in previous development and was too clunky to use locally in functions like `setAll`.
The new `PaletteIndexMap` is also more performant, as it reduces total allocations and exposes several internal functions that reduce double work.

### CODE REVIEW

File: src/main/java/net/minestom/server/instance/palette/Palette.java [hunk 1]
```diff
@@ -30,13 +31,13 @@ public sealed interface Palette permits PaletteImpl {
...
     int BLOCK_PALETTE_MAX_BITS = 8;
-    int BLOCK_PALETTE_DIRECT_BITS = 15;
+    int BLOCK_PALETTE_DIRECT_BITS = MathUtils.bitsToRepresent(Block.statesCount() - 1);
 
     int BIOME_DIMENSION = 4;
     int BIOME_PALETTE_MIN_BITS = 1;
     int BIOME_PALETTE_MAX_BITS = 3;
     @ApiStatus.Internal
-    int BIOME_PALETTE_DIRECT_BITS = 6; // Vary based on biome count, this is just a sensible default
+    int BIOME_PALETTE_DIRECT_BITS = 7; // Vary based on biome count, this is just a sensible default
```
Comment (kermandev) [inline src/main/java/net/minestom/server/instance/palette/Palette.java]: I know these are dynamic, but why dont we set the initial value during startup? Ideally these could be set as late as possible in a class holder idiom.

Likely a biomeDirectBits() static method, with a blockDirectBits() too possibly.

File: src/main/java/net/minestom/server/instance/palette/PaletteImpl.java [hunk 2]
```diff
@@ -17,16 +14,15 @@
 
 final class PaletteImpl implements Palette {
     private static final ThreadLocal<int[]> WRITE_CACHE = ThreadLocal.withInitial(() -> new int[SECTION_BLOCK_COUNT]);
-    final byte dimension, minBitsPerEntry, maxBitsPerEntry, directBits;
+    final byte dimension, minBitsPerEntry, maxBitsPerEntry;
+    byte directBits;
```
Comment (kermandev) [inline src/main/java/net/minestom/server/instance/palette/PaletteImpl.java]: Any major reason why you just dont throw on mismatch for this and keep it final?

File: src/main/java/net/minestom/server/instance/palette/PaletteImpl.java [hunk 3]
```diff
@@ -191,61 +124,38 @@ public void replace(int oldValue, int newValue) {
...
-                valueToPaletteMap.remove(oldValue);
-                valueToPaletteMap.put(newValue, index);
-                // Update count
-                if (newValue == 0) {
-                    this.count -= count; // Replacing with air
-                } else if (oldValue == 0) {
-                    this.count += count; // Replacing air with a block
-                }
+            int oldIndex;
+            int newIndex;
```
Comment (kermandev) [inline src/main/java/net/minestom/server/instance/palette/PaletteImpl.java]: final keyword here please (for both)

File: src/main/java/net/minestom/server/instance/palette/Palette.java [hunk 4]
```diff
@@ -68,24 +69,35 @@ static Palette sized(int dimension, int minBitsPerEntry, int maxBitsPerEntry, in
...
+     * Efficiently fills a cuboid from {@code (x0, y0, z0)} to {@code (x1, y1, z1)} (inclusive).
+     * <p>
+     * If any part of the cuboid lies outside the bounds of palette,
+     * those out-of-bounds positions are ignored.
+     * <p>
+     * Coordinate order does not matter.
+     *
+     * @param value The value to fill with
+     */
+    void fill(int x0, int y0, int z0, int x1, int y1, int z1, int value);
```
Comment (kermandev) [inline src/main/java/net/minestom/server/instance/palette/Palette.java]: add fill(Point, Point, int)

File: src/main/java/net/minestom/server/instance/palette/PaletteImpl.java [hunk 5]
```diff
@@ -260,20 +170,118 @@ public void replace(int x, int y, int z, IntUnaryOperator operator) {
...
+    public void fill(int value) {
+        this.bitsPerEntry = 0;
+        this.count = value;
+        this.values = null;
+        this.paletteIndexMap = null;
+    }
+
+    @Override
+    public void fill(int x0, int y0, int z0, int x1, int y1, int z1, int value) {
+        final int dimensionMinus = dimension - 1;
```
Comment (kermandev) [inline src/main/java/net/minestom/server/instance/palette/PaletteImpl.java:222]: We should not do bounds clamping here, instead we should throw on out of bounds.

File: src/main/java/net/minestom/server/instance/palette/PaletteImpl.java [hunk 6]
```diff
@@ -285,65 +293,31 @@ public void copyFrom(Palette source, int offsetX, int offsetY, int offsetZ) {
...
+        // Calculate the actual copy bounds - only copy what fits within target bounds
+        final int minX = Math.max(0, -offsetX);
+        final int minY = Math.max(0, -offsetY);
+        final int minZ = Math.max(0, -offsetZ);
+        final int maxX = dimension - Math.max(0, offsetX);
+        final int maxY = dimension - Math.max(0, offsetY);
+        final int maxZ = dimension - Math.max(0, offsetZ);
+
+        // Early exit if nothing to copy (offset pushes everything out of bounds)
+        if (minX >= maxX || minY >= maxY || minZ >= maxZ) {
```
Comment (kermandev) [inline src/main/java/net/minestom/server/instance/palette/PaletteImpl.java]: throw instead (I know its breaking, but it doesnt make sense it turn into a noop for out of bounds)

File: src/main/java/net/minestom/server/instance/palette/PaletteIndexMap.java [hunk 7]
```diff
@@ -0,0 +1,204 @@
...
+package net.minestom.server.instance.palette;
+
+import it.unimi.dsi.fastutil.HashCommon;
+import net.minestom.server.utils.MathUtils;
+
+/**
+ * A performant implementation for transforming values into palette indices.
+ * Derived from {@link it.unimi.dsi.fastutil.ints.Int2IntOpenHashMap} and modified to suit our specific needs.
+ */
+@SuppressWarnings({"StatementWithEmptyBody", "UnusedReturnValue"})
```
Comment (kermandev) [inline src/main/java/net/minestom/server/instance/palette/PaletteIndexMap.java]: remove first suppressed

File: src/main/java/net/minestom/server/instance/palette/PaletteIndexMap.java [hunk 8]
```diff
@@ -0,0 +1,204 @@
...
+        PaletteIndexMap c;
+        try {
+            c = (PaletteIndexMap) super.clone();
+        } catch (CloneNotSupportedException cantHappen) {
+            throw new InternalError();
+        }
+        c.containsNullKey = containsNullKey;
+        c.value = value.clone();
+        c.index = index.clone();
+        c.indexToValue = indexToValue.clone();
```
Comment (kermandev) [inline src/main/java/net/minestom/server/instance/palette/PaletteIndexMap.java:198]: missing cloning the rest

File: src/main/java/net/minestom/server/instance/palette/PaletteIndexMap.java [hunk 9]
```diff
@@ -0,0 +1,204 @@
...
+    private int realSize() {
+        return containsNullKey ? size - 1 : size;
+    }
+
+    @Override
+    public PaletteIndexMap clone() {
+        PaletteIndexMap c;
+        try {
+            c = (PaletteIndexMap) super.clone();
+        } catch (CloneNotSupportedException cantHappen) {
```
Comment (kermandev) [inline src/main/java/net/minestom/server/instance/palette/PaletteIndexMap.java]:
```suggestion
        } catch (CloneNotSupportedException _) {
```

File: src/main/java/net/minestom/server/instance/palette/PaletteImpl.java [hunk 10]
```diff
@@ -285,65 +293,31 @@ public void copyFrom(Palette source, int offsetX, int offsetY, int offsetZ) {
...
-                        if (wasAir != isAir) {
-                            this.count += wasAir ? 1 : -1;
-                        }
-                    }
-                }
-            }
+        if (sourcePalette.bitsPerEntry == 0 || sourcePalette.count == 0) {
+            final int dimensionMinus = dimension - 1;
+            fill(offsetX, offsetY, offsetZ,
+                    dimensionMinus + offsetX, dimensionMinus + offsetY, dimensionMinus + offsetZ,
```
Comment (kermandev) [inline src/main/java/net/minestom/server/instance/palette/PaletteImpl.java]: Wouldnt this always throw if its not zero now, Maybe this shouldnt take account the offsetX/Y/Z for the upper bound

File: src/main/java/net/minestom/server/instance/palette/Palette.java [hunk 11]
```diff
@@ -109,8 +121,10 @@ static Palette sized(int dimension, int minBitsPerEntry, int maxBitsPerEntry, in
      */
     void copyFrom(Palette source);
 
+    void load(int[] palette, long[] values);
```
Comment (kermandev) [inline src/main/java/net/minestom/server/instance/palette/Palette.java:157]: javadoc please

File: src/main/java/net/minestom/server/instance/palette/Palette.java [hunk 12]
```diff
@@ -68,24 +69,35 @@ static Palette sized(int dimension, int minBitsPerEntry, int maxBitsPerEntry, in
...
 
     void replaceAll(EntryFunction function);
 
+    void fill(int value);
+
+    /**
+     * Efficiently fills a cuboid from {@code (x0, y0, z0)} to {@code (x1, y1, z1)} (inclusive).
+     * <p>
+     * If any part of the cuboid lies outside the bounds of palette,
+     * those out-of-bounds positions are ignored.
```
Comment (kermandev) [inline src/main/java/net/minestom/server/instance/palette/Palette.java]: throws on any out of bounds positions.

Imagine Palette is an array, if you try do something outside of the bounds of the array, the logic of your program is wrong, especially since access to chunk palettes are synchronized, and if you can avoid accessing them for nops you should.

File: src/main/java/net/minestom/server/instance/palette/Palette.java [hunk 13]
```diff
@@ -231,19 +256,24 @@ static NetworkBuffer.Type<Palette> serializer(int dimension, int minIndirect, in
...
             public void write(NetworkBuffer buffer, Palette palette) {
                 PaletteImpl value = (PaletteImpl) palette;
-                // Temporary fix for biome direct bits depending on the number of registered biomes
-                if (directBits != value.directBits && !value.hasPalette()) {
-                    PaletteImpl tmp = new PaletteImpl((byte) dimension, (byte) minIndirect, (byte) maxIndirect, (byte) directBits);
-                    tmp.setAll(value::get);
-                    value = tmp;
+                Check.argCondition(dimension != value.dimension,
+                        "Palette must be of dimension {0}, got {1}", dimension, value.dimension);
+                if (directBits != value.directBits && value.isDirect()) {
```
Comment (kermandev) [inline src/main/java/net/minestom/server/instance/palette/Palette.java]: Would still prefer the workaround, as it can be alleviated with more general registry initialization changes, Unlikely this is going to be a bottleneck for your server especially if the default bits turns into a class holder idiom which should be lazy enough.

File: src/main/java/net/minestom/server/instance/palette/Palette.java [hunk 14]
```diff
@@ -30,13 +31,13 @@ public sealed interface Palette permits PaletteImpl {
     int BLOCK_DIMENSION = 16;
     int BLOCK_PALETTE_MIN_BITS = 4;
     int BLOCK_PALETTE_MAX_BITS = 8;
-    int BLOCK_PALETTE_DIRECT_BITS = 15;
+    int BLOCK_PALETTE_DIRECT_BITS = MathUtils.bitsToRepresent(Block.statesCount() - 1);
```
Comment (kermandev) [inline src/main/java/net/minestom/server/instance/palette/Palette.java]: Could you mark this as internal, as now its dependent on registry initialization?

**Review by kermandev — COMMENTED** [review]
Thanks for the PR, I do think it does clean it up a bit, but do you have any benchmarks against master?

File: src/main/java/net/minestom/server/instance/palette/Palette.java (hunk 1 above)
Comment (AidanMars2) [reply src/main/java/net/minestom/server/instance/palette/Palette.java]: This would be a good idea, and I tried it, but unfortunately we don't have the biome registry available during testing (it is part of the server process and that is null during the tests as far as I know)

File: src/main/java/net/minestom/server/instance/palette/PaletteImpl.java (hunk 2 above)
Comment (AidanMars2) [reply src/main/java/net/minestom/server/instance/palette/PaletteImpl.java]: The registries are only frozen after the `MinecraftServer#start()` has been called. 

Before that happens, an instance can be created, which creates biome palettes.
Then you can still register new biome types, potentially changing the direct bits, which pre-existing palettes don't know about.

File: src/main/java/net/minestom/server/instance/palette/Palette.java (hunk 1 above)
Comment (kermandev) [reply src/main/java/net/minestom/server/instance/palette/Palette.java]: You just need to add EnvTest to the accompanying tests with a Env paramater, but realistically sections or chunks should be passing down the direct sizes they require

File: src/main/java/net/minestom/server/instance/palette/PaletteImpl.java (hunk 2 above)
Comment (kermandev) [reply src/main/java/net/minestom/server/instance/palette/PaletteImpl.java]: Creating an instance wouldn't initialize palettes, creating chunks would, but I understand your point.

**Comment by AidanMars2** [comment]
> Thanks for the PR, I do think it does clean it up a bit, but do you have any benchmarks against master?

I benchmarked against master and got the following:
`get` and `set` were not significantly affected.
The new `fill` overload is 4x faster than the naive implementation.
`getAll` and `getAllPresent` are now 1.5-2x faster.
`optimize` is now 25% faster.

`setAll` and `replaceAll` are 3x slower than master, using 3-4x less memory and network in exchange.

**Comment by kermandev** [comment]
> > Thanks for the PR, I do think it does clean it up a bit, but do you have any benchmarks against master?
> 
> I benchmarked against master and got the following: `get` and `set` were not significantly affected. The new `fill` overload is 4x faster than the naive implementation. `getAll` and `getAllPresent` are now 1.5-2x faster. `optimize` is now 25% faster.
> 
> `setAll` and `replaceAll` are 3x slower than master, using 3-4x less memory and network in exchange.

Do you have the JMH benchmarks you were using for these tests or are you using the server to test them?

**Comment by AidanMars2** [comment]
https://github.com/AidanMars2/palette-test

note: the claim of less memory and network is based on my understanding of the palette structure

**Comment by kermandev** [comment]
So Ive reread this, i realized it makes no sense that directBits needs to be mutable or is changing at all, its defined as the upper bound so BPE <= directBits, assume its like maxDirectBits.
I think its actually a problem that was overlooked by the current implementations, so if any value is greater the possible max value in maxDirectBits it should fail. Also I think the current tests fail this cause they attempt filling negative values, or values to big to fit into maxDirectBits ever.

Also biomeBits shouldnt change if REGISTRY_UNSAFE_OPS is disabled after startup, but yes there is a window between when biomes are added and possibly instances being created with chunks, so I think the original workaround is more correct, with a possible warning that its having to resize up. (As now it should throw if the value is too big for the max direct bits, which can be diagnosed easier from an initialization order.

Also If either is not a single value:
Optimizing for speed should make BPE = directBits
Optimizing for size should try to shrink BPE as much as possible, even in direct mode.

Also sidenote: Technically since we control the PaletteIndexMap, it KV/I could be optimized in the same packing for space knowing the bounds of key/values. I wouldnt worry about that until a future PR though.

**Comment by AidanMars2** [comment]
> So Ive reread this, i realized it makes no sense that directBits needs to be mutable or is changing at all, its defined as the upper bound so BPE <= directBits, assume its like maxDirectBits.

This assumption is incorrect, as per the [minecraft wiki](https://minecraft.wiki/w/Java_Edition_protocol/Chunk_format#Paletted_Container_structure).

> Values not listed in the following table are rounded upwards to the next one specified, or downwards if larger than the value for Direct. Therefore such values will lead to unexpected results, and should not be used.

So the following needs to hold true exactly `bpe == 0 || (bpe >= minBits && bpe <= maxBits) || bpe == directBits` to be able to serialize without resizing.

**Comment by kermandev** [comment]
> > So Ive reread this, i realized it makes no sense that directBits needs to be mutable or is changing at all, its defined as the upper bound so BPE <= directBits, assume its like maxDirectBits.
> 
> This assumption is incorrect, as per the [minecraft wiki](https://minecraft.wiki/w/Java_Edition_protocol/Chunk_format#Paletted_Container_structure).
> 
> > Values not listed in the following table are rounded upwards to the next one specified, or downwards if larger than the value for Direct. Therefore such values will lead to unexpected results, and should not be used.
> 
> So the following needs to hold true exactly `bpe == 0 || (bpe >= minBits && bpe <= maxBits) || bpe == directBits` to be able to serialize without resizing.

Yes protocol wise it is required directBits are required to be equal from both sides, but form a compactness its possible to compress them smaller than that until writing.

Still we do not need to support directBits resizing, instead only bpe should change, and any value that is greater than the maximum bpe defined in directBits cannot be stored.

**Comment by AidanMars2** [comment]
So I benchmarked the dynamic direct bits and found the following:
`get` and `set` are not measurably affected when not resizing.
1632–3504 bytes can be saved when using lower direct bits (out of 8192 bytes by default).
Serializing becomes 30x slower according to [this benchmark](https://github.com/AidanMars2/palette-test/blob/d2932b3f93b472b6f789c59f993362c145cc106c/src/main/java/com/aidanmars/bench/WriteValuesBench.java), where the values are funneled into a `BlackHole`.

If you decide you want the memory savings in exchange for the performance hit in serialization, that's fine with me. Just let me know.

Side note: we currently cache full chunk packets in a way where all palettes in the chunk are essentially stored twice in memory. I think it's a better optimization to serialize the palettes on the fly.

**Comment by kermandev** [comment]
I think it might be an idea to be more conformant to the protocol with the later option to let users #optimize(SIZE) that performs that direct size optimization. Also with the chunk packet cache, its actually quite important to have this, and users can disable these anyways behind server flags. (Not really documented though!)

**Comment by AidanMars2** [comment]
I tend to agree, but I do want to point out that if someone wants to `#optimze(SIZE)` to improve network performance, it can reduce serialization performance. It might be worth adding `#optimize(NETWORK)` to get around this.

**Comment by kermandev** [comment]
Superseeded by #3324, which does some more changes to palettes. Apologies for not getting this in after a lot of back and forth review.

Some changes can likely be reintroduced after it lands. (Likely around 26.3 minestom)
